src/main/java/com/amazon/ion/impl/macro/Expression.kt

```diff
@@ -0,0 +1,215 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.impl.macro
+
+import com.amazon.ion.*
+import java.math.BigDecimal
+import java.math.BigInteger
+
+/**
+ * In-memory expression model.
+ *
+ * We cannot use [`IonValue`](com.amazon.ion.IonValue) for this because `IonValue` requires references to parent
+ * containers and to an IonSystem which makes it impractical for reading and writing macros definitions. Furthermore,
+ * there is information we need to capture that cannot be expressed in the IonValue model, such as macro invocations
+ * and variable references.
+ *
+ * Template bodies are compiled into a list of expressions, without nesting, for ease and efficiency of evaluating
+ * e-expressions. Because of this, the container types do not have other values nested in them; rather they contain a
+ * range that indicates which of the following expressions are part of that container.
+ */
+sealed interface Expression {
+
+    /** Interface for expressions that "contain" other expressions */
+    sealed interface HasStartAndEnd : Expression {
+        /**
+         * The position of this expression in its containing list.
+         * Child expressions (if any) start at `selfIndex + 1`.
+         */
+        val selfIndex: Int
+        /**
+         * The index of the first child expression (if any).
+         * Always equal to `selfIndex + 1`.
+         */
+        val startInclusive: Int get() = selfIndex + 1
+        /**
+         * The exclusive end of the child expressions (if any).
+         * If there are no child expressions, will be equal to [startInclusive].
+         */
+        val endExclusive: Int
+    }
+
+    /** Marker interface representing expressions that can be present in E-Expressions. */
+    sealed interface EncodingExpression : Expression
```

Comment:
🗺️  The name `EncodingExpression` is a little confusing because it is _not_ an `EExpression` (that's further down), but I haven't been able to come up with a better name yet.

Reply:
> EExpressionBodyExpression

I like it.